docs/proposals/dynamic-kubelet-configuration.md

```diff
@@ -0,0 +1,319 @@
+- The Kubelet's configuration type will no longer align with it's flags; we should add a separate struct that contains the flag variables to prevent them from migrating all over the codebase (similar to work being done in [#32215](https://github.com/kubernetes/kubernetes/issues/32215)). IN PROGRESS: [#40117](https://github.com/kubernetes/kubernetes/pull/40117).
+- We should add more structure to the Kubelet configuration for readability.
+- We should expose Kubelet API groups via groups in `pkg/kubelet/apis` and move the location of the configuration type to this subtree - e.g. to `pkg/kubelet/apis/config`. IN PROGRESS: [#42759](https://github.com/kubernetes/kubernetes/pull/42759).
+- We need to be able to add and remove experimental fields from the `KubeletConfiguration` without having to rev the API version. A simple solution is to just have a string representation of the experimental fields as part of the KubeletConfiguration, that the Kubelet can parse as necessary. THIS OPTION: [#41082](https://github.com/kubernetes/kubernetes/pull/41082). That said, this option is essentially a hack, and it is likely better just to rely on the standard API versioning mechanism. The overhead of up-versioning `pkg/kubelet/apis/config` is low enough to do so, because the former only relates to one component. This is much easier than versioning `pkg/apis/componentconfig`. Eventually, fine-grain field versioning ([#34508](https://github.com/kubernetes/kubernetes/issues/34508)) will solve this entirely, but it may be a while before that work is complete.
+
+
+### Representing and Referencing Configuration
+
+#### Cluster-level object
+- The Kubelet's configuration should be stored as a `JSON` or `YAML` blob under the `kubelet` key in a `ConfigMap` object. This allows the `ConfigMap`'s schema to be extended to other `Node`-level components, e.g. adding a key for `kube-proxy`.
+- The Kubelet's configuration type should be organized in the cluster as a structured monolith. 
+  + *Structured*, so that it is readable.
+  + *Monolithic*, so that all Kubelet parameters roll out to a given `Node` in unison.
+- The `ConfigMap` containing the desired configuration should be specified via the `Node` object corresponding to the Kubelet. The `Node` will have a new `spec` subfield, `config`, which is an `ObjectReference` intended to refer to a `ConfigMap`.
+- The name of the `ConfigMap` containing the desired configuration should be of the form `blah-blah-blah-{hash}`, where `{hash}` is a SHA-1 hash of the `data` field of the `ConfigMap`. The hash will be produced by serializing the `data` to a `JSON` string, and then taking the hash of this string. Depending on ordering guarantees, we may also need to ensure that keys are sorted in the serialization to ensure consistent hashing.
```

Comment:
Yeah fair point, we might as well because it will make the integrity checking better, even though here it's more about preventing accidents than attacks. 

Does anyone know if Git has settled on a new hash yet? A lot of ideas were tossed around in the thread Linus started https://plus.google.com/+LinusTorvalds/posts/7tp2gYWQugL, but I'm having trouble digging up any final decision.

Picking a hash is like only having enough money for one piece of candy in the whole store...

We could additionally mitigate by appending the length of the data to the string that gets hashed, which I believe is one of the things Git does that makes it more robust against collisions.